Bolts/Common/BFTask.m

```diff
@@ -229,84 +236,114 @@ + (instancetype)taskFromExecutor:(BFExecutor *)executor withBlock:(nullable id (
     }];
 }
 
+- (void)commonInit {
+    _state = pending;
+    _synchronizationQueue = dispatch_queue_create("com.bolts.task", DISPATCH_QUEUE_CONCURRENT);
+}
+
 #pragma mark - Custom Setters/Getters
 
 - (nullable id)result {
-    @synchronized(self.lock) {
-        return _result;
-    }
+    __block id result;
+    dispatch_barrier_sync(_synchronizationQueue, ^{
```

Comment:
You only need to use `barrier_sync` for writes - for reads, just use `dispatch_sync`. You'll get read-write lock behavior here, with a slight perf improvement.